### ISSUE TITLE
In xraru, there is 1/50000 chance of reusing :0

### ISSUE THREAD

**Issue by aexaey** [teran-mckinney/raru#1]
If there is already a native X server running on the host, xraru will reuse the same display (:0) with 1/50000 probability. So line 14 should probably read:

```
DISP=$((1 + $(date +%s) % 50000))
```

**Comment by teran-mckinney** [comment]
I thought about that, too. There's actually several other potentially wrong bits in here that make `xraru` quite likely eratic. I guess, though, if `xhost` had set `:0` to be open to all, this would turn into a security vulnerability, and not just a ctrl+c, up arrow, enter style annoyance.

I remember a discussion about rare cases where the logic to test for oddities is more likely to break than the oddity occuring, itself. That said, 1 in 50,000 is not that rare on computer terms. I can add that in.

Thanks!

**Comment by aexaey** [comment]
You are right, in a sane distro, both ```xhost``` and file permissions on ```~/.Xauthority``` file would probably stop unrelated uid from accessing ```:0```. Still, infrequent failures to start would be mildly annoying I would imagine.

Thinking again about collisions with ```:0``` and also collision with another instances of ```xraru```, maybe we should ask X to auto-allocale a free display with ```-displayfd``` option, like suggested here:

http://stackoverflow.com/a/18166600#18166600

i.e. instead of ```DISP=...```/```env ... raru Xvnc ...```, do this:

```
exec 6>/tmp/xraru-$$
env -i PATH=$PATH raru Xvnc -nolisten tcp -geometry 800x600 -depth 24 -nevershared -dontdisconnect -deferupdate 15 -once -interface 127.0.0.1 -localhost -displayfd 6 &
DISP=$(</tmp/xraru-$$)
rm -f /tmp/xraru-$$
```

**Comment by teran-mckinney** [comment]
That is interesting. Much different from the shell I usually write. It would be more predictable.

The main problem I had was with `vncviewer` being picky about the port. For some reason, the display option didn't work for me so I used the port. Ideally, I'd rather avoid localhost TCP altogether and just use a file socket of some sort. Do you know if tha'ts supported? I know that `Xvnc` does support an inetd mode. If vncviewer did the same and could feed them together with a bidirectional pipe, could avoid this entirely.

Could also try things in a loop testing for failures with different random ports each time, until it works. I'd really need to fix the cleanup in that case.

**Comment by teran-mckinney** [comment]
Thought about this a little more:

> `DISP=$((1 + $(date +%s) % 50000))`

This is very order of operations dependent. Modulo is done before the addition. Otherwise, one second earlier we'd have our zero. I tested this and it seems to act as we want it to.

If `:0` is unsecured anyways, couldn't an evil application just talk to `:0`, without being put there, and then decide to become evil? Of course the code would break, too. Probably worth it, but not a panacea.

If `DISP=0`:

1. `Xvnc` fails.
2. `vncviewer` subsequently fails.
3. `$*` runs in `:0` only if the X security settings allow it. I could also abort on failures of `Xvnc` and/or `vncviewer`, but that's trickier to track when backgrounding them. Maybe Python is the way to go down the road, even though we're ultimately managing just shell commands, at least for now.

**Comment by aexaey** [comment]
Making Xvnc listen on a unix socket cound be done by wrapping it with ```socat```:

```
SOCK=$(mktemp -u)
socat unix-l:$SOCK exec:"Xvnc -inetd (...)"
```

However, vncviewer to my best knowledge doesn't support neither unix sockets nor inetd mode. That said, writing a patch to add unix socket support there shouldn't be too hard. Do you think it worth trouble going that route?

Now about vncviewer being picky to display number - it is most likely due to wacky port/display detection algorithm there: as documented in ```vncviewer --help```, connection target written as ```::NUMBER``` is always treated as tcp port number; what isn't documented is in ```:NUMBER``` syntax, NUMBER should be <100 to be treated as display number, otherwise treated as port number too - see here:

https://github.com/TigerVNC/tigervnc/blob/master/common/rfb/Hostname.h#L44

In other words:

```vncviewer :99``` will connect to tcp/127.0.0.1:5999
```vncviewer :100``` will connect to tcp/127.0.0.1:100

Handholding of Xvnc (with ```expect``` for example) to see if it exited with an error and only backgrounding it after a hold-off timer fired and no error was seen, is possible of course, but would be way more verbose and fragile then using display/port auto-assignment with ```-displayfd```.

**Comment by teran-mckinney** [comment]
Ah, that makes so much more sense. Thank you for digging into that.

I realized that the port method now is not so prone to failure. It's binding only on the local socket. So it'd be X in 50,000 chance for self-connections.

It would be nice to have a stripped down Xvnc and vncviewer more meant for this. I can poke at the codebase and see what it'd take to clean them up a bit. Could strip out a lot designing it for only localhost. Performance is actually better than I thought it would be, but no doubt there is more overhead  than is needed.

Would be especially nice to add resizing support with that.

**Comment by aexaey** [comment]
> It would be nice to have a stripped down Xvnc and vncviewer more meant for this. 

That sounds like ```Xnest``` to me.

Regarding resizing, there were some patches floating around to add ```xrandr``` extension to Xvnc, but as far as I can tell, this was never integrated:

http://www.freesoft.org/software/vncrotation/
